errors/error.go

```diff
@@ -0,0 +1,373 @@
+							"start byte=%d, end line=%d, end col=%d, end byte=%d",
+							v.Filename,
+							v.Start.Line, v.Start.Column, v.Start.Byte,
+							v.End.Line, v.End.Column, v.End.Byte),
+					)
+				} else {
+					errParts = append(errParts, v.String())
+				}
+			}
+		case Kind:
+			if v != "" {
+				errParts = append(errParts, string(v))
+			}
+		case string:
+			if v != "" {
+				errParts = append(errParts, v)
+			}
+		case StackMeta:
+			if v != nil {
+				if verbose {
+					errParts = append(errParts,
+						fmt.Sprintf("at stack (name=%q, path=%q, desc=%q)",
+							v.Name(), v.Path(), v.Desc(),
+						))
+				} else {
+					errParts = append(errParts, fmt.Sprintf("at stack %q", v.Path()))
+				}
+			}
+		case error:
+			if v != nil {
+				errParts = append(errParts, v.Error())
+			}
+		case nil:
+			// ignore nil values
+		default:
+			panic(fmt.Errorf("unexpected errors.E type: %+v", arg))
+		}
+	}
+
+	return strings.Join(errParts, separator)
+}
+
+// Error returns the error message.
+func (e *Error) Error() string {
+	return e.error(false)
+}
+
+// Detailed returns a detailed error message.
+func (e *Error) Detailed() string {
+	return e.error(true)
+}
+
+// Is implements errors.Is interface.
+func (e *Error) Is(target error) bool {
+	return Is(e, target)
+}
+
+// IsKind tells if err is of kind k.
+// It returns false if err is nil or not an *errors.Error.
+// It also recursively checks if any underlying error is of kind k.
+func IsKind(err error, k Kind) bool {
+	if err == nil {
+		return false
+	}
+	e, ok := err.(*Error)
+	if !ok {
+		return false
+	}
+	if e.Kind != "" && e.Kind == k {
+		return true
+	}
+	return IsKind(e.Err, k)
+}
+
+// Is tells if err (or any of its underlying errors) matches target.
+// It works with any error but if comparing *Error type it uses the Kind field,
+// otherwise it fallback to standard errors.Is().
+func Is(err, target error) bool {
+	if (err == nil) != (target == nil) {
+		return false
+	}
+	e, ok := err.(*Error)
+	if !ok {
+		return errors.Is(err, target)
+	}
+	t, ok := target.(*Error)
+	if !ok {
+		// target is not *Error so it's already different.
+		// Let's check the underlying error.
+		if e.Err != nil {
+			return Is(e.Err, target)
+		}
+		return false
+	}
+
+	// both are *Error
+
+	if t.Kind != "" && !IsKind(e, t.Kind) {
+		return false
+	}
+
+	if t.Description != "" && !hasDesc(e, t.Description) {
+		return false
+	}
+
+	if t.Stack != nil && !hasSameStack(e, t.Stack) {
+		return false
+	}
+
+	if !t.FileRange.Empty() && e.FileRange != t.FileRange {
+		return false
+	}
+
+	if t.Err != nil {
+		return Is(e.Err, t.Err)
+	}
+
+	return true
+}
+
+func hasDesc(err error, desc string) bool {
+	e, ok := err.(*Error)
+	if !ok {
+		return false
+	}
+
+	if e.Description == desc {
+		return true
+	}
+	if e.Err != nil {
+		return hasDesc(e.Err, desc)
+	}
+	return false
+}
+
+// hasSameStack recursively check if err or any of its underlying errors has the
+// provided stack set.
```

Comment:
```suggestion
// hasSameStack recursively check if err or any of its underlying errors has
// the same stack metadata.
```

Reply:
this message would be accurate if both params were errors and we were looking if they have the same metadata.

Reply:
hmm but what would be a "stack set" here ? I just don't get it :thinking:

Reply:
... has the provided stack set to its Stack field. ?

Reply:
hmm for some reason I was always thinking about "sets"...like a set of stacks :laughing: . Maybe Im just extra stupid..but I was trying to improve this because it is confusing at least to me. But it is probably an isolated dumbness on my part, disregard.